File: core/src/com/gmaniliapp/flappyhermes/state/PlayState.java
package com.gmaniliapp.flappyhermes.state;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.audio.Sound;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.utils.Array;
import com.gmaniliapp.flappyhermes.FlappyHermes;
import com.gmaniliapp.flappyhermes.sprite.Column;
import com.gmaniliapp.flappyhermes.sprite.Ground;
import com.gmaniliapp.flappyhermes.sprite.Sandal;

public class PlayState extends State {

    private static final int COLUMN_SPACING = 125;
    private static final int COLUMN_COUNT = 4;

    private static final int GROUND_COUNT = 2;

    private Texture background;

    private Array<Ground> grounds;

    private Sandal sandal;

    private Array<Column> columns;

    private Sound hit;
    private Sound point;

    protected PlayState(GameStateManager gameStateManager) {
        super(gameStateManager);

        orthographicCamera.setToOrtho(false, FlappyHermes.WIDTH / 2, FlappyHermes.HEIGHT / 2);

        background = new Texture("background.png");

        grounds = new Array<>();
        for (int i = 0; i < GROUND_COUNT; i++) {
            grounds.add(new Ground(
                    orthographicCamera.position.x - orthographicCamera.viewportWidth / 2, i));
        }

        sandal = new Sandal(50, 320);

        columns = new Array<>();

        for (int i = 1; i <= COLUMN_COUNT; i++) {
            columns.add(new Column(i * (COLUMN_SPACING + Column.WIDTH)));
        }

        hit = Gdx.audio.newSound(Gdx.files.internal("hit.ogg"));
        point = Gdx.audio.newSound(Gdx.files.internal("point.ogg"));
    }

    @Override
    public void update(float deltaTime) {
        handleInput();

        for (int i = 0; i < grounds.size; i++) {
            Ground ground = grounds.get(i);

            if (orthographicCamera.position.x - orthographicCamera.viewportWidth / 2 >
                    ground.getPosition().x + ground.getTexture().getWidth()) {
                ground.reposition(
                        ground.getPosition().x + ground.getTexture().getWidth() * GROUND_COUNT);
            }
            if (ground.collides(sandal.getBounds())) {
                hit.play();
                gameStateManager.set(new MenuState(gameStateManager));
            }
        }

        sandal.update(deltaTime);

        orthographicCamera.position.x = sandal.getPosition().x + 80;

        for (int i = 0; i < columns.size; i++) {
            Column column = columns.get(i);

            if (orthographicCamera.position.x - orthographicCamera.viewportWidth / 2 >
                    column.getPositionTop().x + column.getTop().getWidth()) {
                column.reposition(
                        column.getPositionTop().x +
                                ((Column.WIDTH + COLUMN_SPACING) * COLUMN_COUNT));
                point.play();
            }
            if (column.collides(sandal.getBounds())) {
                hit.play();
                gameStateManager
                        .set(new com.gmaniliapp.flappyhermes.state.MenuState(gameStateManager));
            }
        }

        orthographicCamera.update();
    }

    @Override
    public void render(SpriteBatch spriteBatch) {
        spriteBatch.setProjectionMatrix(orthographicCamera.combined);

        spriteBatch.begin();

        spriteBatch.draw(background,
                orthographicCamera.position.x - (orthographicCamera.viewportWidth / 2),
                orthographicCamera.position.y - (orthographicCamera.viewportHeight / 2));

        spriteBatch.draw(sandal.getTexture(), sandal.getPosition().x, sandal.getPosition().y);

        for (Column column : columns) {
            spriteBatch.draw(column.getTop(), column.getPositionTop().x, column.getPositionTop().y);
            spriteBatch
                    .draw(column.getBottom(), column.getPositionBottom().x,
                            column.getPositionBottom().y);
        }

        for (Ground ground : grounds) {
            spriteBatch.draw(ground.getTexture(), ground.getPosition().x, ground.getPosition().y);
        }

        spriteBatch.end();
    }

    @Override
    public void dispose() {
        background.dispose();
        for (Ground ground : grounds) {
            ground.dispose();
        }
        sandal.dispose();
        for (Column column : columns) {
            column.dispose();
        }
        hit.dispose();
        point.dispose();
    }

    @Override
    protected void handleInput() {
        if (Gdx.input.justTouched()) {
            sandal.jump();
        }
    }
}

File: core/src/com/gmaniliapp/flappyhermes/state/PlayState.kt
package com.gmaniliapp.flappyhermes.state

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.audio.Sound
import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.badlogic.gdx.utils.Array
import com.gmaniliapp.flappyhermes.FlappyHermes
import com.gmaniliapp.flappyhermes.sprite.Column
import com.gmaniliapp.flappyhermes.sprite.Ground
import com.gmaniliapp.flappyhermes.sprite.Sandal

class PlayState(gameStateManager: GameStateManager) : State(gameStateManager) {

    private val background: Texture
    private val grounds: Array<Ground>
    private val sandal: Sandal
    private val columns: Array<Column>
    private val hit: Sound
    private val point: Sound

    override fun update(deltaTime: Float) {
        handleInput()
        for (i in 0 until grounds.size) {
            val ground = grounds[i]
            if (orthographicCamera.position.x - orthographicCamera.viewportWidth / 2 >
                    ground.position.x + ground.texture.width) {
                ground.reposition(
                        ground.position.x + ground.texture.width * GROUND_COUNT)
            }
            if (ground.collides(sandal.bounds)) {
                hit.play()
                gameStateManager.set(MenuState(gameStateManager))
            }
        }
        sandal.update(deltaTime)
        orthographicCamera.position.x = sandal.position.x + 80
        for (i in 0 until columns.size) {
            val column = columns[i]
            if (orthographicCamera.position.x - orthographicCamera.viewportWidth / 2 >
                    column.positionTop.x + column.top.width) {
                column.reposition(
                        column.positionTop.x +
                                (Column.WIDTH + COLUMN_SPACING) * COLUMN_COUNT)
                point.play()
            }
            if (column.collides(sandal.bounds)) {
                hit.play()
                gameStateManager
                        .set(MenuState(gameStateManager))
            }
        }
        orthographicCamera.update()
    }

    override fun render(spriteBatch: SpriteBatch) {
        spriteBatch.projectionMatrix = orthographicCamera.combined
        spriteBatch.begin()
        spriteBatch.draw(background,
                orthographicCamera.position.x - orthographicCamera.viewportWidth / 2,
                orthographicCamera.position.y - orthographicCamera.viewportHeight / 2)
        spriteBatch.draw(sandal.texture, sandal.position.x, sandal.position.y)
        for (column in columns) {
            spriteBatch.draw(column.top, column.positionTop.x, column.positionTop.y)
            spriteBatch
                    .draw(column.bottom, column.positionBottom.x,
                            column.positionBottom.y)
        }
        for (ground in grounds) {
            spriteBatch.draw(ground.texture, ground.position.x, ground.position.y)
        }
        spriteBatch.end()
    }

    override fun dispose() {
        background.dispose()
        for (ground in grounds) {
            ground.dispose()
        }
        sandal.dispose()
        for (column in columns) {
            column.dispose()
        }
        hit.dispose()
        point.dispose()
    }

    override fun handleInput() {
        if (Gdx.input.justTouched()) {
            sandal.jump()
        }
    }

    companion object {
        private const val COLUMN_SPACING = 125
        private const val COLUMN_COUNT = 4
        private const val GROUND_COUNT = 2
    }

    init {
        orthographicCamera.setToOrtho(false, FlappyHermes.WIDTH / 2f, FlappyHermes.HEIGHT / 2f)
        background = Texture("background.png")
        grounds = Array()
        for (i in 0 until GROUND_COUNT) {
            grounds.add(Ground(
                    orthographicCamera.position.x - orthographicCamera.viewportWidth / 2, i))
        }
        sandal = Sandal(50, 320)
        columns = Array()
        for (i in 1..COLUMN_COUNT) {
            columns.add(Column((i * (COLUMN_SPACING + Column.WIDTH)).toFloat()))
        }
        hit = Gdx.audio.newSound(Gdx.files.internal("hit.ogg"))
        point = Gdx.audio.newSound(Gdx.files.internal("point.ogg"))
    }
}